// tdl/dist/prompt-stubs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPhoneNumber = exports.getPassword = exports.getName = exports.getEmailCode = exports.getEmailAddress = exports.getAuthCode = exports.confirmOnAnotherDevice = void 0;

const getPhoneNumber = () => {
  throw new Error('Default getPhoneNumber is not supported');
};

exports.getPhoneNumber = getPhoneNumber;

const getEmailAddress = () => {
  throw new Error('Default getEmailAddress is not supported');
};

exports.getEmailAddress = getEmailAddress;

const getEmailCode = () => {
  throw new Error('Default getEmailCode is not supported');
};

exports.getEmailCode = getEmailCode;

const confirmOnAnotherDevice = () => {
  throw new Error('Default confirmOnAnotherDevice is not supported');
};

exports.confirmOnAnotherDevice = confirmOnAnotherDevice;

const getAuthCode = () => {
  throw new Error('Default getAuthCode is not supported');
};

exports.getAuthCode = getAuthCode;

const getPassword = () => {
  throw new Error('Default getPassword is not supported');
};

exports.getPassword = getPassword;

const getName = () => {
  throw new Error('Default getName is not supported');
};

exports.getName = getName;
